refined/src/main/scala/coulomb/ops/algebra/refined/all.scala

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2022 Erik Erlandson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package coulomb.ops.algebra.refined
+
+import algebra.ring.AdditiveSemigroup
+
+import eu.timepit.refined.*
+import eu.timepit.refined.api.*
+import eu.timepit.refined.numeric.*
+
+object all:
+    inline given ctx_AdditiveSemigroup_Refined[V](using
+        alg: AdditiveSemigroup[V],
+        vld: Validate[V, Positive]
+    ): AdditiveSemigroup[Refined[V, Positive]] =
+        new AdditiveSemigroup[Refined[V, Positive]]:
+            def plus(x: Refined[V, Positive], y: Refined[V, Positive]): Refined[V, Positive] =
+                refineV[Positive].unsafeFrom(alg.plus(x.value, y.value))
```

Comment:
what happens if this addition overflows?

Reply:
It would be an error, but I haven't worked out a way around it. Technically I think it's type unsound, since if it did overflow it could result in a state not reflected by the type.

Reply:
Unless I can check with the validator "manually"

Reply:
The only truly type sound semantic would be to have the resulting value be the `Either` expression, which seems awkward to me, however it would work.

Reply:
Well, it "works" in the sense that any type can be a Quantity value, but the algebra typeclass requires the original V type.

Reply:
You might have a look at this:
https://github.com/gemini-hlsw/refined-algebra/blob/97c3510db3912a0dd57679f11a311644af7c61f8/refined-algebra/src/main/scala/refined/algebra.scala#L13

Reply:
`unsafeFrom` throws a runtime exception if it overflows, or the result otherwise violates the predicate.  I suppose it is a matter of philosophy whether it's "better" to error out, or not error out but silently return an incorrect result (i.e. max-int).  Neither is great, although my intuition is that if I'm getting overflows, I'd want to know it via a runtime error, instead of getting a result that doesn't reflect the actual sum.
 
```scala
scala> refineV[Positive].unsafeFrom(2000000000)
val res3: Int Refined eu.timepit.refined.numeric.Positive = 2000000000
                                                                                                                                                           
scala> refineV[Positive].unsafeFrom(2000000000 + 2000000000)
java.lang.IllegalArgumentException: Predicate failed: (-294967296 > 0).
  at eu.timepit.refined.internal.RefinePartiallyApplied.unsafeFrom$$anonfun$1(RefinePartiallyApplied.scala:21)
  at scala.util.Either.fold(Either.scala:190)
  at eu.timepit.refined.internal.RefinePartiallyApplied.unsafeFrom(RefinePartiallyApplied.scala:21)
  ... 66 elided
```

Reply:
At any rate, the current behavior is NOT unsound, at least in the sense that it won't silently return a Refined with a value that violates the predicate. It will either return a correct value or error out at runtime.

Reply:
> At any rate, the current behavior is NOT unsound

But note that it is unlawful.

Reply:
Also probably depends on whether you are semantically trying to represent a modular ring (as in that example) or an unbounded additive semigroup.

Reply:
On the bright side, `AdditiveSemigroup` for `+`  will admit either my algebra definitions or a user's preferred definition, whichever they want to import.

Reply:
FWIW, one of my design philosophies has been that the "out-of-box" definitions provided by coulomb deliberately do not try to obscure the "native" behaviors of numeric types.  So integer addition can overflow.  Floating points do whatever complicated native floating point behaviors, etc.

Users may very reasonably wish to define better semantics, but the definitions of "better" have a lot of potential variation, and I decided it was clarifying to me to avoid that kind of potential for policy explosion.

Reply:
Sure. FWIW, the design philosophy of algebra `AdditiveSemigroup` and friends is that implementing instances are lawful. Those _are_ the semantics.